arbnode/tx_pre_checker.go

```diff
@@ -6,36 +6,67 @@ package arbnode
 import (
 	"context"
 	"fmt"
+	"time"
 
+	"github.com/ethereum/go-ethereum/arbitrum_types"
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core"
 	"github.com/ethereum/go-ethereum/core/state"
 	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/metrics"
 	"github.com/ethereum/go-ethereum/params"
 	"github.com/offchainlabs/nitro/arbos/arbosState"
 	"github.com/offchainlabs/nitro/arbos/l1pricing"
 	"github.com/offchainlabs/nitro/util/arbmath"
+	"github.com/pkg/errors"
+	flag "github.com/spf13/pflag"
 )
 
+var (
+	conditionalTxRejectedByTxPreCheckerCurrentStateCounter = metrics.NewRegisteredCounter("arb/txprechecker/condtionaltx/currentstate/rejected", nil)
+	conditionalTxAcceptedByTxPreCheckerCurrentStateCounter = metrics.NewRegisteredCounter("arb/txprechecker/condtionaltx/currentstate/accepted", nil)
+	conditionalTxRejectedByTxPreCheckerOldStateCounter     = metrics.NewRegisteredCounter("arb/txprechecker/condtionaltx/oldstate/rejected", nil)
+	conditionalTxAcceptedByTxPreCheckerOldStateCounter     = metrics.NewRegisteredCounter("arb/txprechecker/condtionaltx/oldstate/accepted", nil)
+)
+
+const TxPreCheckerStrictnessNone uint = 0
+const TxPreCheckerStrictnessAlwaysCompatible uint = 10
+const TxPreCheckerStrictnessLikelyCompatible uint = 20
+const TxPreCheckerStrictnessFullValidation uint = 30
+
+type TxPreCheckerConfig struct {
+	Strictness       uint  `koanf:"strictness" reload:"hot"`
+	RequiredStateAge int64 `koanf:"required-state-age" reload:"hot"`
+}
+
+type TxPreCheckerConfigFetcher func() *TxPreCheckerConfig
+
+var DefaultTxPreCheckerConfig = TxPreCheckerConfig{
+	Strictness:       TxPreCheckerStrictnessNone,
+	RequiredStateAge: 1,
+}
+
+func TxPreCheckerConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.Uint(prefix+".strictness", DefaultTxPreCheckerConfig.Strictness, "how strict to be when checking txs before forwarding them. 0 = accept anything, "+
+		"10 = should never reject anything that'd succeed, 20 = likely won't reject anything that'd succeed, "+
+		"30 = full validation which may reject txs that would succeed")
+	f.Int64(prefix+".required-state-age", DefaultTxPreCheckerConfig.RequiredStateAge, "how long ago should the storage conditions from eth_SendRawTransactionConditional be true, 0 = don't check old state")
```

Comment:
Hmm.. It's a bit unfortunate that this can't be fractional. To help with that, let's add a `required-state-max-blocks` which is the maximum number of blocks we'll look back while looking for the block e.g. 1 second ago. If we set it to 2 for instance, it will keep going back until either it's found a block 1 second ago or it's gone back 2 blocks. 0 would mean current functionality of go back however far until we find a sufficiently old block.